crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs

```diff
@@ -0,0 +1,138 @@
+
+        if old_version < 1 {
+            create_stores_for_v1(evt.db())?;
+        }
+
+        if old_version < 2 {
+            create_stores_for_v2(evt.db())?;
+        }
+
+        if old_version < 3 {
+            create_stores_for_v3(evt.db())?;
+        }
+
+        if old_version < 4 {
+            create_stores_for_v4(evt.db())?;
+        }
+
+        if old_version < 5 {
+            create_stores_for_v5(evt.db())?;
+        }
+
+        info!(old_version, new_version, "IndexeddbCryptoStore upgrade complete");
+        Ok(())
+    }));
+
+    db_req.await
+}
+
+fn create_stores_for_v1(db: &IdbDatabase) -> Result<(), DomException> {
+    db.create_object_store(keys::CORE)?;
+    db.create_object_store(keys::SESSION)?;
+
+    db.create_object_store(keys::INBOUND_GROUP_SESSIONS)?;
+    db.create_object_store(keys::OUTBOUND_GROUP_SESSIONS)?;
+    db.create_object_store(keys::TRACKED_USERS)?;
+    db.create_object_store(keys::OLM_HASHES)?;
+    db.create_object_store(keys::DEVICES)?;
+
+    db.create_object_store(keys::IDENTITIES)?;
+    db.create_object_store(keys::BACKUP_KEYS)?;
+
+    Ok(())
+}
+
+fn create_stores_for_v2(db: &IdbDatabase) -> Result<(), DomException> {
+    // We changed how we store inbound group sessions, the key used to
+    // be a tuple of `(room_id, sender_key, session_id)` now it's a
+    // tuple of `(room_id, session_id)`
+    //
+    // Let's just drop the whole object store.
+    db.delete_object_store(keys::INBOUND_GROUP_SESSIONS)?;
+    db.create_object_store(keys::INBOUND_GROUP_SESSIONS)?;
+
+    db.create_object_store(keys::ROOM_SETTINGS)?;
+
+    Ok(())
+}
+
+fn create_stores_for_v3(db: &IdbDatabase) -> Result<(), DomException> {
+    // We changed the way we store outbound session.
+    // ShareInfo changed from a struct to an enum with struct variant.
+    // Let's just discard the existing outbounds
+    db.delete_object_store(keys::OUTBOUND_GROUP_SESSIONS)?;
+    db.create_object_store(keys::OUTBOUND_GROUP_SESSIONS)?;
+
+    // Support for MSC2399 withheld codes
+    db.create_object_store(keys::DIRECT_WITHHELD_INFO)?;
+
+    Ok(())
+}
+
+fn create_stores_for_v4(db: &IdbDatabase) -> Result<(), DomException> {
+    db.create_object_store(keys::SECRETS_INBOX)?;
+    Ok(())
+}
+
+fn create_stores_for_v5(db: &IdbDatabase) -> Result<(), DomException> {
+    // Create a new store for outgoing secret requests
+    let object_store = db.create_object_store(keys::GOSSIP_REQUESTS)?;
+
+    let mut params = IdbIndexParameters::new();
+    params.unique(false);
+    object_store.create_index_with_params(
+        keys::GOSSIP_REQUESTS_UNSENT_INDEX,
+        &IdbKeyPath::str("unsent"),
+        &params,
+    )?;
+
+    let mut params = IdbIndexParameters::new();
+    params.unique(true);
+    object_store.create_index_with_params(
+        keys::GOSSIP_REQUESTS_BY_INFO_INDEX,
+        &IdbKeyPath::str("info"),
+        &params,
+    )?;
+
+    if db.object_store_names().any(|n| n == "outgoing_secret_requests") {
```

Comment:
I found the previous check with respect to the previous version number more obvious, but I think it should be equivalent, assuming atomic operations.